File: code/Game.py
from Player import Player
from Board import Board
from Parser import Parser
from Dice import Dice

class Game:

    def __init__(self, faces, players, board_size, snakes, ladders):

        self.snakes = {}
        self.ladders = {}
        self.board_size = board_size
        self.dice = Dice(faces)

        for src, dest in snakes:
            self.snakes[src] = dest

        for src, dest in ladders:
            self.ladders[src] = dest

        self.board = Board(self.board_size,self.snakes, self.ladders)
        self.players = [Player(name, self.board) for name in players]
    
    def play(self):
        while(1):
            for player in self.players:
                from_position = player.position
                dice_value = self.dice.roll()
                player.move(dice_value)
                to_position = player.position
                print(player.name, "rolled a", dice_value,  "and moved from", from_position, "to", to_position)
                if player.position == self.board_size:
                    return player.name + " wins the game"

parser = Parser()

game = Game(parser.faces, parser.players, parser.board_size, parser.snakes, parser.ladders)
print(game.play())
    

File: src/com/vineetvdubey/snakeladder/game/Game.java
package com.vineetvdubey.snakeladder.game;

import com.vineetvdubey.snakeladder.model.Board;
import com.vineetvdubey.snakeladder.model.Dice;
import com.vineetvdubey.snakeladder.model.Player;

import java.util.Queue;

public class Game {

    private final Board board;
    private final Queue<Player> players;

    public Game(Board board, Queue<Player> players) {
        this.board = board;
        this.players = players;
    }

    public void play() {
        log("\nStarting a new game of Snakes and Ladders!\n");
        Dice dice = board.getDice();
        int finishPosition = 0;
        while (players.size() > 1) {
            Player currentPlayer = players.remove();
            int diceValue = dice.roll();
            int currPosition = currentPlayer.getPosition();
            int newPosition = board.calculateNewPosition(currPosition, diceValue);
            currentPlayer.setPosition(newPosition);
            logRollEvent(currentPlayer.getName(), diceValue, currPosition, newPosition);
            if (newPosition == board.getBoardSize()) {
                finishPosition += 1;
                logFinishEvent(currentPlayer.getName(), finishPosition);
            } else {
                players.add(currentPlayer);
            }
        }
        log("\nThanks for playing!");
    }

    private void logRollEvent(String name, int diceValue, int currPosition, int newPosition) {
        log(name + " rolled a " + diceValue + " and moved from " + currPosition + " to " + newPosition);
    }

    private void logFinishEvent(String name, int finishPosition) {
        log(name + " finished the game (rank " + finishPosition + ")");
    }

    private void log(String message) {
        System.out.println(message);
    }
}
